File: Prototype/Prototype01.js
// Classe Pessoa
class Pessoa {
    constructor(id, nome, idade) {
        this.id = id;
        this.nome = nome;
        this.idade = idade;
    }

    // Método clone
    clone() {
        return new Pessoa(this.id, this.nome, this.idade);
    }
}

// Classe GerenciaPessoa
class GerenciaPessoa {
    constructor() {
        this.pessoas = {};
    }

    // Adiciona nova pessoa
    addPessoa(id, nome, idade) {
        const pessoa = new Pessoa(id, nome, idade);
        this.pessoas[id] = pessoa;
    }

    // Retorna clone da pessoa
    getPessoaById(id) {
        const pessoaOriginal = this.pessoas[id];

        if (pessoaOriginal) {
            return pessoaOriginal.clone();
        } else {
            return null;
        }
    }
}

// Criando uma instância de GerenciaPessoa:
const gerencia = new GerenciaPessoa();

// Adicionando Pessoas
gerencia.addPessoa(1, 'a', 32);
gerencia.addPessoa(2, 'b', 12);
gerencia.addPessoa(3, 'c', 64);

// Clona pessoa 1 e modifica nome
const pessoaClone1 = gerencia.getPessoaById(1);
pessoaClone1.nome = 'd'; 

console.log(gerencia.getPessoaById(1));
console.log("-----Pessoa Clone-------"); 
console.log(pessoaClone1);
console.log("------------------------");
console.log(gerencia.getPessoaById(2)); 
console.log("------------------------");
console.log(gerencia.getPessoaById(3)); 
console.log("------------------------");

File: Prototype/Prototype01.py
import copy

class Pessoa: 
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade
    
    def __str__(self):
        return f'Nome: {self.nome}, Idade: {self.idade}'
    
    def clone(self):
        return copy.copy(self)

class GerenciaPessoa:
    def __init__(self):
        self.pessoas = {}

    def addPessoa(self, nome, idade, id):
        pessoa = Pessoa(nome, idade)
        self.pessoas[id] = pessoa

    def getPessoa(self, id):
        return self.pessoas[id].clone()
    
# Criação da gerência
gerencia = GerenciaPessoa()

# Adicionando pessoas
gerencia.addPessoa('a', 32, 1)
gerencia.addPessoa('b', 12, 2)
gerencia.addPessoa('c', 64, 3)

# Clonando pessoas
pessoa1 = gerencia.getPessoa(1)
pessoa2 = gerencia.getPessoa(2)

# Alterando valores das clones
pessoa1.idade = 18
pessoa2.nome = "EWEE"
pessoa2.idade = 55

# Impressão das pessoas originais na gerência
print(gerencia.getPessoa(1))
print("------------------------")
print(gerencia.getPessoa(2))
print("------------------------")
print(gerencia.getPessoa(3))
print("------------------------")

# Impressão das clones alteradas
print(pessoa1)
print(pessoa2)